collect_app/src/test/java/org/odk/collect/android/widgets/UrlWidgetTest.java

```diff
@@ -98,26 +106,23 @@ public void clickingButton_callsCorrectMethods() {
     }
 
     @Test
-    public void onLongPressButton_doesNotCallOpenUri() {
+    public void clickingButtonForLong_callsLongClickListener() {
         UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
         Button urlButton = widget.findViewById(R.id.url_button);
         urlButton.performLongClick();
 
-        verify(customTabHelper, never()).bindCustomTabsService(widget.getContext(), null);
-        verify(customTabHelper, never()).openUri(widget.getContext(), Uri.parse("blah"));
+        assertThat(listener.onLongClick(urlButton), equalTo(true));
```

Comment:
I think a better way to test this is to pass a `mock(Listener.class)` to the Widget's `OnLongClickListener` and then check that it is called (using `when`) after you call `urlButton.performLongClick()`. Does that make sense? That would make sure the listener the URL button has and the listener the widget has are definitely the same

Reply:
This is much better! But sorry I lead you down a slightly wrong path there. I think instead of using `when` to set up a result and then verifying it later (lines 108 and 111) you could just do `verify(listener).onLongClick(urlButton)`. Does that make more sense?

Reply:
I wrote the test like this initially:
`@Test
    public void clickingButtonForLong_callsLongClickListener() {
        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
        Button urlButton = widget.findViewById(R.id.url_button);

        widget.setOnLongClickListener(listener);
        urlButton.performLongClick();
        verify(listener).onLongClick(urlButton);
    }`

But this test does not pass.

Reply:
Right but I think that's a correctly failing test: `URLWidget#setOnLongClickListener` doesn't do anything at the moment so I'd expect that test to be red. If you try long clicking on the URL button in the app at the moment right now you'll see it doesn't show the "Edit Prompt" pop up.

Reply:
It says `Wanted but not invoked` and `Actually, there were zero interactions with this mock.` I do not understand.

Reply:
Right that means that the `onLongClick` on the listener was never called. It might be worth reading some docs at https://site.mockito.org/ if you're confused by this message. If you look at `URLWidget#setOnLongClickListener` you'll see it's empty so that test should fail in the way you're seeing. You have a red test that you can now make green!

Reply:
Hi! I'm sorry. I didn't refresh the page and so kept adding comments without reading your second last comment.

Reply:
@SaumiaSinghal hah I've made exactly the same mistake before!

Reply:
:disappointed_relieved: